### ISSUE TITLE
[Linux only] kustomization.yaml not found if it's a symlink

### ISSUE THREAD

**Issue by SanjayVas** [kubernetes-sigs/kustomize#4420]
**Describe the bug**

`kustomize build` returns the following error when run in a directory where a `kustomization.yaml` file exists, but is a symlink rather than a regular file.

```
Error: unable to find one of 'kustomization.yaml', 'kustomization.yml' or 'Kustomization' ' in directory
```

It works as expected when `kustomization.yaml` with the same contents is a regular file.

**Kustomize version**

v4.4.1

**Platform**

Linux

**Comment by KnVerey** [comment]
/retitle [Linux only] kustomization.yaml not found if it's a symlink

This issue doesn't exist on OSX, but I was able to reproduce on Linux.

/triage accepted

**Comment by SanjayVas** [comment]
I think there may be other cases that aren't handled by changing the load restrictor, e.g. referencing other files in the kustomization.yaml that are symlinks.

I believe this prevents running kustomize from Bazel, as Bazel relies heavily on symlinking to put both generated and hand-written source files in the "same location".

**Comment by kurt-google** [comment]
Just to confirm it does prevent you from using kustomize in Bazel without workarounds. `--load-restrictor LoadRestrictionsNone` resolves this issue in bazel.

**Comment by SanjayVas** [comment]
> `--load-restrictor LoadRestrictionsNone` resolves this issue in bazel.

I don't think this is true. There's some other issues with symlink resolution. If you have a mix of generated and source files, things get screwy since some of those get symlinked. Kustomize appears to expect all paths to be relative to the resolved location rather than the symlinked one.

The only workaround I've found is to copy all the files to a single tree (in Bazel, this means tarring them up and extracting them later).

**Comment by kurt-google** [comment]
Sorry I did not to meant to say it resolves all possible issues. In our case we do not have generated files, so the only issue is this issue of kustomization.yaml not being found if its a symlink. Constructing a source tree for the rule does sound like the safest bet.

Any efforts towards making kustomize work better in bazel builds are much appreciated! Thanks for the tip on the workaround.

**Comment by HariSekhon** [comment]
Not supporting symlinks has been a hindrance for quite a long time... and it's still the case on the very latest version of kustomize v.4.5.5.

Symlinks are a core part of unix administration, use cases include: 

- testing different adjacent versions of `kustomization.yaml` by replacing symlinks
- a repo full of different kustomizations that I would like to just symlink to a working directory and execute, eg:

https://github.com/HariSekhon/Kubernetes-configs

Workarounds such as `cp` instead of `ln` are not ideal, because you are duplicating contents and then have to keep both copies up to date - this is exactly the sort of thing symlinks were designed to solve.

I'm hoping it's a fairly easy file existence call check to fix this?

**Comment by annasong20** [comment]
Until `kustomize build` behavior is changed, [`kustomize localize`](https://github.com/kubernetes-sigs/kustomize/blob/master/proposals/22-04-localize-command.md) will follow the `kustomize build` standard and not accept files that are symlinks to locations outside the kustomization root.

**Comment by sanmai-NL** [comment]
@SanjayVas could you please reopen?

**Comment by SanjayVas** [comment]
> @SanjayVas could you please reopen?

/reopen
/remove-lifecycle rotten

**Comment by sanmai-NL** [comment]
Not stale

**Comment by akirchhoff-modular** [comment]
I poked around the code, and after applying this patch, Kustomize seems more than happy to accept symlinks:

```patch
diff --git a/kyaml/filesys/fsondisk.go b/kyaml/filesys/fsondisk.go
index 10ce83f12..2f1cd9657 100644
--- a/kyaml/filesys/fsondisk.go
+++ b/kyaml/filesys/fsondisk.go
@@ -55,11 +55,7 @@ func (x fsOnDisk) CleanedAbs(
                return "", "", fmt.Errorf(
                        "abs path error on '%s' : %v", path, err)
        }
-       deLinked, err := filepath.EvalSymlinks(absRoot)
-       if err != nil {
-               return "", "", fmt.Errorf(
-                       "evalsymlink failure on '%s' : %w", path, err)
-       }
+       deLinked := absRoot
        if x.IsDir(deLinked) {
                return ConfirmedDir(deLinked), "", nil
        }
```

However, this is subverting the definition of `ConfirmedDir`:

```go
// ConfirmedDir is a clean, absolute, delinkified path
// that was confirmed to point to an existing directory.
type ConfirmedDir string
```

With this change, `ConfirmedDir` is definitely not delinkified. Frankly, I don't think it needs to be -- The commit that introduced it never explained why it needed to be, and things seem to work even when it isn't. But I am not sure I'd be willing to put this up as a PR, given this.

**Comment by fentas** [comment]
@KnVerey Could we reopen this and discuss the solution of @akirchhoff-modular .
It seems simple enough, but I assume it would be related to security, and if we could go with this, if `--load-restrictor LoadRestrictionsNone` is set.
